gcp_variant_transforms/libs/annotation/annotation_parser.py

```diff
@@ -301,6 +301,79 @@ def _find_alt_index_by_allele_num(self, annotation_map):
       raise InvalidAlleleNumValue(e)
 
 
+class AnnotationStrBuilder(object):
+  """The class for reconstructing annotation str."""
+
+  def __init__(self, annotation_id_to_annotation_names):
+    # type: (Dict[str, List[str]]) -> None
+    """Initializes an object of `AnnotationStrBuilder`.
+
+    Args:
+      annotation_id_to_annotation_names: A map where the key is the annotation
+        id (e.g., `CSQ`) and the value is a list of annotation names (e.g.,
+        ['allele', 'Consequence', 'IMPACT', 'SYMBOL']). The annotation str
+        (e.g., 'A|upstream_gene_variant|MODIFIER|PSMF1|||||') is reconstructed
+        in the same order as the annotation names.
+    """
+    self._annotation_id_to_annotation_names = annotation_id_to_annotation_names
+
+  def reconstruct_annotation_str(self, annotation_id, annotation_maps):
+    # type: (str, List[Dict[str, Any]]) -> List[str]
+    """Returns annotation strings reconstructed from `annotation_map`.
+
+    Notice that the returned value is a list of annotation string since the
+    annotation record is a repeated field. The annotation str (e.g.,
+    'A|upstream_gene_variant|MODIFIER|PSMF1|||||') is
+    reconstructed in the same order as `annotation_names`.
+
+    Args:
+      annotation_id: The annotation id of the `annotation_maps`. Example: 'CSQ'.
+      annotation_maps: A list of annotation map to be reconstructed to
+        annotation str. Example:
+        [{'allele': 'G',
+          'Consequence': 'upstream_gene_variant',
+          'AF': ''
+          'IMPACT': 'MODIFIER'},
+         {'allele': 'G',
+          'Consequence': 'upstream_gene_variant',
+          'AF': '0.1',
+          'IMPACT': ''}
+        ]
+
+    Returns:
+      A list of annotation string reconstructed from `annotation_maps`.
+        Example:
+          ['G|upstream_gene_variant||MODIFIER',
+           'G|upstream_gene_variant|0.1|']
+    """
+    if not self.is_valid_annotation_id(annotation_id):
+      raise ValueError('No annotation names for {} are defined. The '
+                       'annotation string cannot be reconstructed since the '
+                       'order is not provided.'.format(annotation_id))
+    annotation_strs = []
+    for annotation_map in annotation_maps:
+      annotation_values = []
+      for annotation_name in self._annotation_id_to_annotation_names.get(
+          annotation_id):
+        value = annotation_map.get(annotation_name)
+        if value is None:
+          annotation_values.append(_MISSING_ANNOTATION_FIELD_VALUE)
+        else:
+          annotation_values.append(str(value))
+      annotation_strs.append('|'.join(annotation_values))
```

Comment:
I am pretty sure I wrote this comment in my previous review too but not sure where it was dropped in the process:

Have you considered making this into a generator and `yield` here instead of returning the list? While in general I prefer lists as returned value (as it makes code more readable) when the returned value is big enough, then it's worth considering generators as it avoids a big memory copy of the returned value and may have non-insignificant performance consequences.

Reply:
@bashir2 Thanks for the comment! The returned/yielded value is extended to another list, which is saved as the value of the map `info[KEY]` (e.g. KEY = 'CSQ'). Do you mean use info[KEY]=generator rather than info[KEY] = list[str]? I feel it is not feasible to use a generator as the value here.

Reply:
Thanks Bashir for sharing your thoughts offline. PTAL :)

Reply:
@allieychen Under what situation do you think can generators be an issue?  What do you think the approach should be in that situation?